packages/lib-classifier/src/store/FeedbackStore.js

```diff
@@ -1,45 +1,76 @@
 import { autorun } from 'mobx'
-import { addDisposer, getRoot, onAction, types } from 'mobx-state-tree'
+import { addDisposer, addMiddleware, getRoot, onAction, types } from 'mobx-state-tree'
+import { flatten } from 'lodash'
 
 import helpers from './feedback/helpers'
 import strategies from './feedback/strategies'
 
 const FeedbackStore = types
   .model('FeedbackStore', {
     isActive: types.optional(types.boolean, false),
-    rules: types.map(types.frozen({}))
+    rules: types.map(types.frozen({})),
+    showModal: types.optional(types.boolean, false),
+    allowSubjectAdvance: types.optional(types.boolean, false)
   })
+  .views(self => ({
+    get hideSubjectViewer () {
+      return flatten(Array.from(self.rules.values()))
+        .some(rule => rule.hideSubjectViewer)
+    },
+    get messages () {
+      return flatten(Array.from(self.rules.values()))
+        .map(rule => {
+          if (rule.success && rule.successEnabled) {
+            return rule.successMessage
+          } else if (!rule.success && rule.failureEnabled) {
+            return rule.failureMessage
+          }
+        })
+    }
+  }))
   .actions(self => {
     function afterAttach () {
-      createSubjectObserver()
       createClassificationObserver()
-    }
-
-    function createSubjectObserver () {
-      const subjectDisposer = autorun(() => {
-        const subject = getRoot(self).subjects.active
-        if (subject) {
-          self.reset()
-          self.createRules(subject)
-        }
-      })
-      addDisposer(self, subjectDisposer)
+      createSubjectMiddleware()
+      createSubjectObserver()
     }
 
     function createClassificationObserver () {
       const classificationDisposer = autorun(() => {
         onAction(getRoot(self).classifications, (call) => {
           if (call.name === 'completeClassification') {
             const annotations = getRoot(self).classifications.currentAnnotations
-            for (const value of annotations.values()) {
-              self.update(value)
-            }
+            annotations.forEach(annotation => self.update(annotation))
           }
         })
       })
       addDisposer(self, classificationDisposer)
     }
 
+    function createSubjectMiddleware () {
+      const subjectMiddleware = autorun(() => {
+        addMiddleware(getRoot(self).subjects, (call, next, abort) => {
+          if (call.name === 'advance' && self.isActive && self.messages.length && !self.allowSubjectAdvance) {
+            self.showFeedback()
+            return abort()
```

Comment:
Ideally, this should pause the action, show feedback (or whichever intervention is happening) then continue the paused action. That will become more important when we implement interventions in the new classifier.

Reply:
something like
```javascript
yield intervention()
next(call)
```
will pause the call while `intervention()` executes then resume the interrupted call, if I'm reading the middleware docs correctly, and if generators can be used here.

Reply:
Ah I think this is exactly what I was missing. As calling the advance from `hideFeedback` didn't feel right. I thought I might have to use volatile/callback like in the classifications store, but this with yield looks much better. Thanks!

Reply:
Argh I still can't get this to work without `abort()` or without calling the subject advance from `hideFeedback`.

I tried `yield` without luck, but likely on me. I'm not sure how to use `yield` effectively here, but probably missing something basic. Or maybe we need to create a promise that's resolved by closing the modal? I thought we were trying to avoid a promise here, but could be totally wrong on that.

I tried saving `next` and `call` to the store state, utilizing [`volatile`](https://mobx-state-tree.gitbook.io/docs/concepts/volatile-state) for `next`, but would get an error while it was essentially waiting for the modal to close, as next or abort needs to be called within the middleware.

Reply:
I added links to the MST docs on middleware and async actions to the initial comment for easy reference.

Reply:
I've found [the MDN docs for yield](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield) useful but, yes, in my toy example `intevention()` would be a Promise that resolves when the intervention is finished.

Reply:
Are there reasons not to use a promise here, resolved with the feedback modal close? 
My hesitation to use a promise might be misplaced from issues that I think we've had with ModalFormDialog.

Reply:
Beats me. Is there a discussion somewhere that I could look at to catch up on whether we should use promises?

Reply:
I don't know where you got to with `yield` but I'm seeing this error when I try to use a generator as middleware: `Neither the next() nor the abort() callback within the middleware flowSpawner for the action: "reset" on the node: RootStore was invoked.`

Reply:
Yes that's the same issue I was having, I think it needs next or abort called within the middleware pretty much right away.

Reply:
But the errors happening on `self.reset()` and the default behaviour is `next(call)`. I'll have another look.

Reply:
I should have read [the middleware page](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/middleware.md) more closely:
>Note: It is important to invoke next(call) or abort(value) synchronously.